Homework4/DiagonalMatrixT/Matrix.cs

```diff
@@ -0,0 +1,92 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace DiagonalMatrixT
+{
+
+    public class Matrix<T> : IEnumerable<T>
+    {
+        public event EventHandler<MatrixEventArgs<T>> ElementChanged;
+
+        public int Size
+        {
+            get; private set;
+        }
+        public T? OldValue { get; set; }
+        private T[] DiagonalElements { get; set; }
+
+        public Matrix(int size)
+        {
+            if (size < 0)
+            {
+                throw new ArgumentException("Matrix size cant be less than 0");
+            }
+
+            Size = size;
+
+            DiagonalElements = new T[Size];
+        }
+
+        public T this[int i, int j]
+        {
+            get
+            {
+                if (i < 0 || j < 0 || i >= Size || j >= Size)
+                {
+                    throw new ArgumentException("Out of range SquareMatrix indexes");
+                }
+                else if (i == j)
+                {
+                    return DiagonalElements[i];
+                }
+                else
+                {
+                    return default(T);
+                }
+            }
+            set
+            {
+
+                if (i < 0 || j < 0 || i >= Size || j >= Size)
+                {
+                    throw new ArgumentException("Out of range SquareMatrix indexes");
+                }
+                else if (i == j)
+
+                {
+                    if (!DiagonalElements[i].Equals(value))
+                    {
+                        OldValue = DiagonalElements[i];
```

Comment:
keep oldValue in a variable instead of public property
```suggestion
                        var oldValue = DiagonalElements[i];
```

Reply:
I can't I use this value in line 79